README.md

```diff
@@ -198,289 +189,492 @@ e.increment();  // logs 1
+### The element being wrapped: the first parameter
+
+The first parameter, of what the decorator is wrapping, depends on what is being decorated:
+- In a method, init-method, getter or setter decorator: the relevant function object
+- In a class decorator: the class
+- In a field: An object with two properties
+    - `get`: A function which takes no arguments, expected to be called with a receiver which is the appropriate object, returning the underlying value.
+    - `set`: A function which takes a single argument (the new value), expected to be called with a receiver which is the object being set, expected to return `undefined`.
+
+### The context object: the second parameter
+
+The context object--the object passed as the second argument to the decorator--contains the following properties:
+- `kind`: One of
+    - `"class"`
+    - `"method"`
+    - `"init-method"`
+    - `"getter"`
+    - `"setter"`
+    - `"field"``
+- `name`:
+    - Public field or method: the `name` is the String or Symbol property key.
+    - Private field or method: missing (could be provided as some representation of the private name, in a follow-on proposal) 
+    - Class: missing
+- `isStatic`:
+    - Static field or method: `true`
+    - Instance field or method: `false`
+    - Class: missing
+
+The "target" (constructor or prototype) is not passed to field or method decorators, as it has not yet been built when the decorator runs.
+
+### The return value
+
+The return value is interpreted based on the type of decorator. The return value is expected as follows:
+- Class: A new class
+- Method, getter or setter: A new function
+- field: An object with three properties (each individually optional):
+    - `get`: A function of the same form as the `get` property of the first argument
+    - `set`: Ditto, for `set`
+    - `initialize`: A called with the same arguments as `set`, which returns a value which is used for the initializing set of the variable. This is called when initially setting the underlying storage based on the field initializer or method definition. This method shouldn't call the `set` input, as that would trigger an error. If `initialize` isn't provided, `set` is not called, and the underlying storage is written directly. This way, `set` can count on the field already existing, and doesn't need to separately track that.
+- Init method: An object with the properties
+    - `method`: A function to replace the method
+    - `initialize`: A function with no arguments, whose return value is ignored, which is called with the newly constructed object as the receiver.
+
+## 3. Applying decorators
+
+Decorators are applied after all decorators have been called. The intermediate steps of the decorator application algorithm are not observable--the newly constructed class is not made available until after all method and non-static field decorators have been applied.
+
+The class decorator is called only after all method and field decorators are called and applied.
```

Comment:
Hmm, do I understand correctly: 

* I have a class decorator like `@defineElement`:
```js
function defineElement(name, options) {
  console.log(1); // Decorator initialization
  return klass => {
    console.log(2); // Decorator application
    customElements.define(name, klass, options);
    return klass;
  }
}
```
* I also have a decorator for some methods:
```js
function logged(f) {
  console.log(3); // Decorator initialization

  function wrapped(...args) {
    console.log(`starting ${name} with arguments ${args.join(", ")}`);
    const ret = f.call(this, ...args);
    console.log(`ending ${name}`);
    return ret;
  }

  return wrapped;
}
```

Am I correct thinking that for the code:
```js
@defineElement('x-foo')
class Foo extends HTMLElement {
  @logged
  render() {}
}
```
the console output will be the following?
```
1
3
2
```
Or the class decorator will be called at the very end? 
```
3
1
2
```

Reply:
1 3 2

Reply:
Great, thanks! That would allow me to initialize internal storage in the class decorator and use it for all class member decorators.

Reply:
Huh, where would you put the storage?

Reply:
Hmm, thinking about this again, I understand that I don't have any info about class in the member decorators. Probably, I should use annotations instead then...